lib/teiserver/account/libs/user_totp_lib.ex

```diff
@@ -0,0 +1,215 @@
+
+  defp get_user_totp(%User{id: user_id}) do
+    case Repo.get_by(TOTP, user_id: user_id) do
+      nil ->
+        {:inactive, nil}
+
+      totp ->
+        {:active, totp}
+    end
+  end
+
+  @spec get_user_totp_status(User.t()) :: :active | :inactive
+  def get_user_totp_status(%User{} = user) do
+    {status, _totp} = get_user_totp(user)
+    status
+  end
+
+  @spec get_account_locked(User.t()) :: boolean
+  def get_account_locked(%User{} = user) do
+    {_status, totp} = get_user_totp(user)
+
+    cond do
+      is_nil(totp) ->
+        false
+
+      totp.wrong_otp < 5 ->
+        false
+
+      totp.wrong_otp >= 5 ->
+        true
+    end
+  end
+
+  @spec get_or_generate_secret(User.t()) :: {:new | :existing, binary()}
+  def get_or_generate_secret(%User{id: _user_id} = user) do
+    case get_user_totp(user) do
+      {:inactive, nil} ->
+        {:new, NimbleTOTP.secret()}
+
+      {:active, totp} ->
+        {:existing, totp.secret}
+    end
+  end
+
+  @spec get_user_secret(User.t()) :: binary | nil
+  def get_user_secret(%User{id: _user_id} = user) do
+    case get_user_totp(user) do
+      {:inactive, nil} ->
+        :inactive
+
+      {:active, totp} ->
+        totp.secret
+    end
+  end
+
+  @spec get_last_used_otp(User.t()) :: :inactive | NaiveDateTime.t()
+  def get_last_used_otp(%User{id: _user_id} = user) do
+    case get_user_totp(user) do
+      {:inactive, nil} ->
+        :inactive
+
+      {:active, totp} ->
+        totp.last_used
+    end
+  end
+
+  @spec set_totp(User.t() | TOTP.t(), map()) ::
+          {:ok, TOTP.t()} | {:error, Ecto.Changeset.t()}
+  defp set_totp(%User{id: nil} = _user, _attrs) do
+    changeset = TOTP.changeset(%TOTP{}, %{})
+    {:error, %{changeset | errors: [user: {"user must be persisted", []}]}}
+  end
+
+  defp set_totp(%User{} = user, attrs) do
+    case get_user_totp(user) do
+      {:inactive, nil} ->
+        TOTP.changeset(%TOTP{}, attrs)
+        |> Repo.insert()
+
+      {:active, totp} ->
+        set_totp(totp, attrs)
+    end
+  end
+
+  defp set_totp(%TOTP{} = totp, attrs) do
+    totp
+    |> TOTP.changeset(attrs)
+    |> Repo.update()
+  end
+
+  @spec set_secret(User.t(), binary) :: {:ok, TOTP.t()} | {:error, Ecto.Changeset.t()}
+  def set_secret(%User{} = user, secret) do
+    set_totp(user, %{user_id: user.id, secret: secret})
+  end
+
+  @spec set_last_used(User.t(), NaiveDateTime.t()) ::
+          {:ok, TOTP.t()} | {:error, Ecto.Changeset.t()} | {:inactive, User.t()}
+  def set_last_used(%User{} = user, last_used) do
+    if get_user_totp_status(user) == :active do
+      set_totp(user, %{user_id: user.id, last_used: last_used})
+    else
+      {:inactive, user}
+    end
+  end
+
+  @spec disable_totp(User.t()) :: {:ok, TOTP.t() | nil} | {:error, Ecto.Changeset.t()}
+  def disable_totp(%User{id: _user_id} = user) do
+    case get_user_totp(user) do
+      {:active, totp} ->
+        Repo.delete(totp)
+
+      {:inactive, _} ->
+        {:ok, nil}
+    end
+  end
```

Comment:
It seems unnecessary that all of these functions are taking the whole user struct as parameter when only the user id is actually needed.
This means that the callers which already have user id now have to unnecessarily get the whole user to use these functions.

While it doesn't appear to be the problem in your current code, because you are using the whole `user` in other places, IMO it's still an unnecessary requirement which could lead to unnecessary queries in the future.